astropy/tests/helper.py

```diff
@@ -409,7 +410,10 @@ def generic_recursive_equality_test(a, b, class_history):
     Check if the attributes of a and b are equal. Then,
     check if the attributes of the attributes are equal.
     """
-    dict_a = a.__getstate__() if hasattr(a, '__getstate__') else a.__dict__
+    if hasattr(a, '__getstate__'):
+        dict_a = a.__getstate__(a) if inspect.isclass(a) else a.__getstate__()
+    else:
+        dict_a = a.__dict__
```

Comment:
It would be good to add a comment here to clarify that the `hasattr` check is only needed for Python < 3.11.

Reply:
If Python 3.11 implemented `__getstate__()` in `object` then `hasattr(a, "__getstate__")` will be true for any `a`. The check is therefore only needed until 3.11 is the minimum supported Python version. A comment would record that it's safe to simplify once support for older versions of Python is dropped.

Reply:
> some class defining their own custom `__getstate__`

Though obviously defining one without parameters would break stuff in 3.11+...

Reply:
OK, I misremembered, thought it was only implemented for `class`.
But in that case it is certainly more likely to get an `a` with a different `__getstate__` signature – e.g. in the original example, `astropy.coordinates.builtin_frames.ICRS.__getstate__` has one required argument, while `astropy.coordinates.builtin_frames.ICRS().__getstate__` takes no arguments.

Reply:
All I'm saying is that the `hasattr` check will serve no purpose after support for Python 3.10 is dropped, so it should be recorded that it can then be safely removed.

Reply:
I agree about adding some sort of Python version check to let us know in the future when the `else` can be removed.

Reply:
Technically something like `if PYTHON_LT_3_11 or hasattr(a, '__getstate__'):` would not change the condition at all, so maybe comments should suffice now?

Reply:
Even if you do a `# PYTHON_LT_3_11` around that line would be fine by me, but let's see what @eerovaher say.

Reply:
Using `PYTHON_LT_3_11` would be even better than only a comment because the flag is actual code and would prevent these temporary lines from becoming permanent.

I have stumbled upon quite a few instances in `astropy` where compatibility code like this has long overstayed its welcome because no one had noticed an adjacent comment, so getting this right here does make a difference.

Reply:
You mean for using some auto-refactoring tool?

Reply:
More like `grep` heh heh

Reply:
Well, `grep` should find the comment as well; humans might not ;-)

Reply:
Oh, well, if we do it Eero's way, the code will actually error when we remove that flag from compat, so maybe that is better.

Reply:
I pushed a commit that would hopefully settle this debate once and for all.